### PR TITLE
Sharded fields dump

### PR DESCRIPTION
**Pull request by kkg4theweb**
Add support for dumping (fields::dump) and loading fields (fields::load). This along with structure::{dump,load} this will allow saving the simulation state at any arbitrary time-step and then restoring from this saved state to continue the simulation from that point.

This PR also adds a basic C++ test dump/load structure and fields. It also moves the existing structure dump/load tests into its own python test (test_dump_load.py) and adds tests for dumping/loading fields also.

### CODE REVIEW

**Comment by oskooi** [comment]
In `test_dump_load.py`, after replacing the line:
```
sim.run(mp.at_every(5, get_field_point), until=1)
```
with
```
sim.init_sim()
```
The subtest `test_load_dump_fields` fails because of  the`Al` (aluminum) material used in one of the two `Block` objects of the `geometry`:
```
        geometry = [mp.Block(material=Al, center=mp.Vector3(), size=one_by_one),
                    mp.Block(material=mp.Medium(epsilon=13), center=mp.Vector3(1), size=one_by_one)]
```
The reason for this test failure is simple: there is a `polarization_state` object of the `fields` class which is *not* being dumped and loaded along with the other field objects as it needs to be:
https://github.com/NanoComp/meep/blob/cdae1a830fd5e25b359a64676317d63e6a59f390/src/meep.hpp#L1466 

As a verification that this is indeed the cause of the failing test, replace `material=Al` with e.g. `material=mp.Medium(index=3.2)` in the `geometry` definition above and `test_dump_load.py` will then pass.

The slight challenge though for dumping/loading the `polarization_state` is that it is a typedef:

https://github.com/NanoComp/meep/blob/cdae1a830fd5e25b359a64676317d63e6a59f390/src/meep.hpp#L1424-L1429

in which the `susceptibility` object is a generic class:

https://github.com/NanoComp/meep/blob/cdae1a830fd5e25b359a64676317d63e6a59f390/src/meep.hpp#L73-L75

Dispersive materials such as `Al` are `lorentzian_susceptibility` objects which derived from `susceptibility`:

https://github.com/NanoComp/meep/blob/cdae1a830fd5e25b359a64676317d63e6a59f390/src/meep.hpp#L236-L240

There are other derived classes of `susceptibility` besides `lorentzian_susceptibility` such as `multilevel_susceptibility`, `gyrotropic_susceptibility`, etc but for now we can just support checkpointing `lorentzian_susceptiblity` objects since they are more common than the other varieties.

File: src/fields_dump.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+   *
+   * When 'single_parallel_file' is true, we are creating a single block of data
+   * for ALL chunks (that are merged using MPI). Otherwise, we are just
+   * making a copy of just the chunks that are ours.
+   */
+  int my_num_chunks = 0;
+  for (int i = 0; i < num_chunks; i++) {
+    my_num_chunks += (single_parallel_file || chunks[i]->is_mine());
+  }
+  size_t num_f_size = my_num_chunks * NUM_FIELD_COMPONENTS * 2;
```
[the remark this answers is not in the document]
Comment (stevengj) [reply src/fields_dump.cpp:48]: If `!is_real` then the imaginary pointers are NULL and so it should do the right thing, I think?

[the remark this answers is not in the document]
Comment (kkg4theweb) [reply src/fields_dump.cpp:48]: A few lines below you can see that we dump (or load) only does it for the non-null complex field component.

**Comment by stevengj** [comment]
The fact that you have to use `init_sim`, and that restoring the checkpoint does not work if you run for a nonzero amount of time first, seems to indicate that there is some state we are forgetting about.  It would be good to chase this down.

File: python/tests/test_dump_load.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                            chunk_layout=sim1)
+        # Just restore structure first.
+        sim.load(dump_dirname, load_structure=True, load_fields=False, single_parallel_file=single_parallel_file)
+        field_points = {}
+
+        def get_field_point(sim):
+            p = sim.get_field_point(mp.Ez, sample_point)
+            field_points[sim.meep_time()] = p.real
+
+        sim.init_sim()
```
Comment (oskooi) [inline python/tests/test_dump_load.py:156]: Removing this line with `sim.init_sim()` causes the subtest `test_load_dump_fields` to fail with a segmentation fault after the previously saved fields are loaded from the HDF5 file and `sim.run` is invoked.

Here is the output from gdb:
```
Running test_load_dump_fields
-----------
Initializing structure...
Halving computational cell along direction y
time for choose_chunkdivision = 0.000307183 s
Working in 2D dimensions.
Computational cell is 5 x 5 x 0 with resolution 50
     block, center = (0,0,0)
          size (1,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (10.24,10.24,10.24)
     block, center = (1,0,0)
          size (1,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (13,13,13)
time for set_epsilon = 0.0874734 s
-----------
run 0 finished at t = 15.0 (1500 timesteps)
creating epsilon from file "/tmp/meepO1ahOS/test_load_dump_fields/structure.h5" (1)...
Dumped structure to file: /tmp/meepO1ahOS/test_load_dump_fields/structure.h5 (True)
creating fields output file "/tmp/meepO1ahOS/test_load_dump_fields/fields.h5" (1)...
Fields State:
  a = 50, dt = 0.01
  m = 0, beta = 0
  t = 1500, phasein_time = 0, is_real = 1

  num_chunks = 6 (shared=1)
Dumped fields to file: /tmp/meepO1ahOS/test_load_dump_fields/fields.h5 (True)
run 1 finished at t = 20.0 (2000 timesteps)
-----------
Initializing structure...
Halving computational cell along direction y
time for choose_chunkdivision = 0.000286404 s
Working in 2D dimensions.
Computational cell is 5 x 5 x 0 with resolution 50
time for set_epsilon = 0.0759436 s
-----------
reading epsilon from file "/tmp/meepO1ahOS/test_load_dump_fields/structure.h5" (1)...
Loaded structure from file: /tmp/meepO1ahOS/test_load_dump_fields/structure.h5 (True)
reading fields from file "/tmp/meepO1ahOS/test_load_dump_fields/fields.h5" (1)...
Fields State:
  a = 50, dt = 0.01
  m = 0, beta = 0
  t = 1500, phasein_time = 0, is_real = 0

  num_chunks = 6 (shared=1)
Loaded fields from file: /tmp/meepO1ahOS/test_load_dump_fields/fields.h5 (True)
on time step 1500 (time=15), 7.26514 s/step

Thread 1 "python3.5" received signal SIGSEGV, Segmentation fault.
0x00007fffc8e6a8e0 in meep::step_curl_stride1 (f=0x55555681f940, c=meep::Bx, g1=0x0, g2=0x0, s1=0, s2=0, gv=..., 
    is=..., ie=..., dtdx=-0.5, dsig=meep::Y, sig=0x5555567505b0, kap=0x5555567503e0, siginv=0x555556750210, 
    fu=0x0, dsigu=meep::NO_DIRECTION, sigu=0x0, kapu=0x0, siginvu=0x0, dt=0.01, cnd=0x0, cndinv=0x0, fcnd=0x0)
    at step_generic_stride1.cpp:192
192	            f[i] = ((kap[k] - sig[k]) * f[i] - dtdx * (g1[i + s1] - g1[i])) * siginv[k];
```
The two terms in this update equation which are causing the segfault are `g1[i + s1]` and `g1[i]`. In `_load_dump_fields`, the `g1` array corresponds to the `Ez` field.

As additional info, the backtrace is:
```
#0  0x00007fffc8e6a8e0 in meep::step_curl_stride1 (f=0x55555681f940, c=meep::Bx, g1=0x0, g2=0x0, s1=0, s2=0, 
    gv=..., is=..., ie=..., dtdx=-0.5, dsig=meep::Y, sig=0x5555567505b0, kap=0x5555567503e0, 
    siginv=0x555556750210, fu=0x0, dsigu=meep::NO_DIRECTION, sigu=0x0, kapu=0x0, siginvu=0x0, dt=0.01, cnd=0x0, 
    cndinv=0x0, fcnd=0x0) at step_generic_stride1.cpp:192
#1  0x00007fffc8e03647 in meep::fields_chunk::step_db (this=0x5555567c4e10, ft=meep::B_stuff) at step_db.cpp:119
#2  0x00007fffc8e02961 in meep::fields::step_db (this=0x555556757c10, ft=meep::B_stuff) at step_db.cpp:36
#3  0x00007fffc8e0022e in meep::fields::step (this=0x555556757c10) at step.cpp:67
#4  0x00007fffc92ac56c in _wrap_fields_step (args=0x7fffc98d3f98) at meep-python.cxx:82395
```

**Comment by oskooi** [comment]
I think I know why `init_sim()` is necessary. To understand this involves going through the sequence of allocating the fields arrays that Meep is using. The important point is that the field allocation (i.e., for `Ez` which is the cause of the segmentation fault reported above) during `init_sim()` actually happens when calling `add_sources`:

https://github.com/NanoComp/meep/blob/f5b60e5b8e9decfd65166dc1f3f82947aecb259a/python/simulation.py#L1985

`add_sources` in turn calls the member function `require_source_components()` of the `fields` class:

https://github.com/NanoComp/meep/blob/f5b60e5b8e9decfd65166dc1f3f82947aecb259a/python/simulation.py#L2445-L2450

`require_source_components` then calls the lower-level function `_require_component` for each of the fields required by the particular source component. In the unit test `_load_dump_fields`, the source component is `Ez` (which is actually an electric current J<sub>z</sub>) and therefore the components in the 2d cell are `Ez`, `Bx`, and `By`. For μ=1 (non-magnetic materials), Meep uses Hx=Bx and Hy=By to avoid unnecessary duplication. Note that in 2d there are two orthogonal sets of polarizations (Ex,Ey,Hz vs. Hx,Hy,Ez) and typically only one set is used in the simulation.

https://github.com/NanoComp/meep/blob/f5b60e5b8e9decfd65166dc1f3f82947aecb259a/src/fields.cpp#L495-L517

Finally, `_require_component` calls `alloc_f` to allocate the fields:

https://github.com/NanoComp/meep/blob/f5b60e5b8e9decfd65166dc1f3f82947aecb259a/src/fields.cpp#L541-L561

https://github.com/NanoComp/meep/blob/f5b60e5b8e9decfd65166dc1f3f82947aecb259a/src/fields.cpp#L465-L493

What needs to happen then when dumping and loading the fields is that only those components generated by the particular source need to be saved and allocated. This means that we should save the *source component* as part of the checkpoint and use it to allocate the correct set of fields when loading the fields similar to how it is already set up.

**Review by stevengj — APPROVED** [review]

**Comment by oskooi** [comment]
It seems there is a memory leak in both `structure::load` and `fields:load` (both introduced in this PR) which is triggered by the C++ unit test `dump_test.cpp` as reported by the [logs of the address sanitizer](https://gist.github.com/ahoenselaar/13d0b764bb38bbab8b579ed414c49592#file-asan-log-for-mep-1e43687-L413).

**Pull request merged**